File: 2_Add_Two_Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        ans = ListNode(0)
        copy = ans
        carry = 0
        
        while l1 or l2:
            if l1 and l2:
                n = (l1.val + l2.val + carry) % 10
                carry = (l1.val + l2.val + carry) // 10
                l1 = l1.next
                l2 = l2.next
            elif l1:
                n = (l1.val + carry) % 10
                carry = (l1.val + carry) // 10
                l1 = l1.next            
            elif l2:
                n = (l2.val + carry) % 10
                carry = (l2.val + carry) // 10
                l2 = l2.next
            
            copy.val = n
            
            if l1 or l2:
                copy.next = ListNode(0)
                copy = copy.next
            
        if carry == 1:
            copy.next = ListNode(1)
        
        return ans
            
            
        

File: swift/2_add_two_numbers.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */

class Solution {
  /// addTwoNumbers returns the sum of two numbers as a linked list.
  func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
    let dummyHead: ListNode = ListNode(0)
    var curNode: ListNode = dummyHead

    var nodeOne: ListNode? = l1
    var nodeTwo: ListNode? = l2

    var carryUp: Int = 0

    while nodeOne != nil || nodeTwo != nil {
      let newNode: ListNode = ListNode(0)

      let nodeOneVal = nodeOne?.val ?? 0
      let nodeTwoVal = nodeTwo?.val ?? 0

      newNode.val = (nodeOneVal + nodeTwoVal + carryUp) % 10
      carryUp = (nodeOneVal + nodeTwoVal + carryUp) / 10

      curNode.next = newNode
      curNode = newNode

      nodeOne = nodeOne?.next
      nodeTwo = nodeTwo?.next
    }

    if carryUp == 1 {
      curNode.next = ListNode(1)
    }

    return dummyHead.next
  }
}
